Review download code

sechkova opened the pull request:
Fixes #1399

**Description of the changes being introduced by the pull request**:

_This PR builds on top of #1408 and will remain a draft until its merge._

- simplify download code, remove average download speed check
- wrap the download functionality in `FileDownload` class
- simplify `RequestsFetcher.fetch()`
- add type annotations

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
Remove average download speed check

The average download speed check does not achieve its
purpose of protecting against slow retrieval attacks.
Such protection is highly dependent on the networking
stack which could be user-provided and not under
tuf's control.
The slow retrieval attack protection has been removed
from the specification for similar reasons.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Remove strict_required_length option

Remove the strict length check in download.py and
check only if an upper bowndary is exceeded.

Exact length is not known for some metadata files
during download. The check makes the code unnecessarily
complicated and is better suited for the later metadata
verification step.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Reduce download logging

Remove logging when an exception is raised.
Reduce logging level to debug.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Implement download_file as contextmanager

It allows us to remove the disable of pylint consider-using-with
warning and does not require any manual file closing.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Simplify RequestsFetcher.fetch()

Make the internal function chunks() a separate
private method of Reqests fetcher.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Remove format checks

Remove format checking of download_file() parameters.
The function is internal and receives trusted input from
updater.py.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Add type annotations

Add missing type annotations in download.py, fetcher.py
and requests_fetcher.py
Update docstrings to match the new style.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
Reduce logging in requests_fetcher

Remove repetitive debug logging.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

jku commented:
Looks great at first glance. 

This maybe a dumb question but is there something that prevents us from putting the new FileDownloader functionality directly into FetcherInterface?

Are the RequestsFetcher improvements dependent on the other changes or just included in same PR for convenience?

sechkova commented:
> This maybe a dumb question but is there something that prevents us from putting the new FileDownloader functionality directly into FetcherInterface?
> 
Only me not thinking enough about it ... :roll_eyes: 

> Are the RequestsFetcher improvements dependent on the other changes or just included in same PR for convenience?

No, they are not dependent, I just consider fetcher as part of the download code.

I do like the idea of simplifying further and keeping only a single download/fetcher module instead of the two. How should we do it? Build on top of this PR, a separate PR or I can rewrite these commits here and directly move download.py functions into FetcherInterface?

jku commented:
> I do like the idea of simplifying further and keeping only a single download/fetcher module instead of the two. How should we do it? Build on top of this PR, a separate PR or I can rewrite these commits here and directly move download.py functions into FetcherInterface?

I think merging a PR that adds a new object would be weird if we intend to remove it soon in a followup PR... otherwise whatever works for you I think

sechkova committed:
Move download.py functions to FetcherInterface

Make download_file and download_bytes functions part of
the FetcherInterface class. Remove download.py module.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Rewrote part of the commits here. Now the simplified download functions are made part of `FetcherInterface` and the `download.py` module is removed.

sechkova commented:
I don't know what is blocking GitHub actions here?

jku commented:
> I don't know what is blocking GitHub actions here?

quite strange. It never actually started the action (https://github.com/theupdateframework/tuf/actions) so I can't even restart it.

Looks really nice btw... Net result of removing 200 lines and ending up with a cleaner implementation is not too bad.

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
+        Yields:
+          A file object that points to the contents of 'url'.
+        """
+        # 'url.replace('\\', '/')' is needed for compatibility with
+        # Windows-based systems, because they might use back-slashes in place
+        # of forward-slashes. This converts it to the common format.
+        # unquote() replaces %xx escapes in a url with their single-character
+        # equivalent. A back-slash may beencoded as %5c in the url, which
+        # should also be replaced with a forward slash.
+        url = parse.unquote(url).replace("\\", "/")
```
let's get rid of this: this component should expect to get a real URL and should not do this.

Do we need a new issue to handle possible issues in higher level code?

sechkova replied:
Opened a new issue although there are several ones related to paths and urls in Updater already: https://github.com/theupdateframework/tuf/issues/1483

jku replied:
oh you might be right about multiple bugs -- I probably file one every time I see this code

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
         raise NotImplementedError  # pragma: no cover
+
+    @contextmanager
+    def download_file(self, url: str, required_length: int) -> Iterator[IO]:
```
required_length is actually more like max_length, right?

sechkova replied:
yes, it is the upper boundary.

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
+              lengths while downloading the file.
+
+        Yields:
+          A file object that points to the contents of 'url'.
```
I wonder if it's worth mentioning that it's a TemporaryFile?

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -40,3 +48,51 @@ def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
+            temp_file.seek(0)
+            yield temp_file
+
+    def download_bytes(self, url: str, required_length: int) -> bytes:
```
(same comment on required_length naming)

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -39,3 +48,51 @@ def fetch(self, url, required_length):
+              the file or an upper boundary.
+
+        Raises:
+          DownloadLengthMismatchError: a mismatch of observed vs expected
+              lengths while downloading the file.
```
```suggestion
          DownloadLengthMismatchError: File length was larger than <name of argument>
```

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -86,58 +87,62 @@ def fetch(self, url, required_length):
+                # large file in one shot. Before beginning the round, sleep
+                # (if set) for a short amount of time so that the CPU is not
+                # hogged in the while loop.
+                if settings.SLEEP_BEFORE_ROUND:
+                    time.sleep(settings.SLEEP_BEFORE_ROUND)
+
+                read_amount = min(
+                    settings.CHUNK_SIZE,
+                    required_length - bytes_received,
+                )
```
let's get rid of settings: 
* CHUNK_SIZE: this is a RequestsFetcher internal detail. make it an instance attribute (maybe with a smaller default than the current huge number)?
* SLEEP_BEFORE_ROUND: if this truly is an issue we should fix the issue -- there is no reason for us to be running a hot while loop for downloading things, that's just crazy.  #338 documents that that is indeed how the code used to work... but that was a long time ago, and the code looked entirely different. I'd say let's spend a bit of time making sure the loop is sane (and doesn't hog cpu) and let's remove SLEEP_BEFORE_ROUND

jku replied:
for reference, pip and requests use default chunk size of 1024*10

jku replied:
this is alread being handled 1470 so maybe forget the comment here -- we should get rid of sleep before round but doesn't need to happen here.

sechkova replied:
I opened https://github.com/theupdateframework/tuf/issues/1484 for it and we can continue discussing settings in the #1470 thread.

jku reviewed (commented):
I like it a lot: this makes download code much easier to understand (and I agree with the decisions made)! Left several comments but the only one that I think needs to be fixed here is the argument name: _required_length_ does not seem to represent what it is.

Feel free to file issues for the other things

sechkova committed:
Replace required_length with max_length

Rename required_length argument in fetcher to
max_length to better match its purpose.
Docstring improvements.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
I think all suggestions are added with 959ab5a.

sechkova commented:
Rebased on latest changes in develop.

jku commented on tuf/ngclient/_internal/requests_fetcher.py:
```diff
@@ -53,7 +53,7 @@ def __init__(self):
         self.sleep_before_round: Optional[int] = None
 
-    def fetch(self, url, required_length):
+    def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
```
I guess this is more `max_length` than `required_length` as well?

the docstring could be better as well

jku commented on tuf/ngclient/fetcher.py:
```diff
@@ -20,7 +29,7 @@ class FetcherInterface:
 
     @abc.abstractmethod
-    def fetch(self, url, required_length):
+    def fetch(self, url: str, required_length: int) -> Iterator[bytes]:
```
maybe should be max_length (also docstring could be better)?

This is sort of API already but the change maybe worth it as required_length really is misleading...

jku reviewed (approved):
Left comments/question about naming of the argument to Fetcher.fetch() as well... but I'm approving as is, this is a very good improvement on download code IMO

sechkova commented:
Amended the last commit 0eff004 with the suggestion.

MVrachev reviewed (approved):
Great job!
I liked all of the simplifications.
It's easier to read now.
It looks good to me as well.
